### PR TITLE
Add support for codecov.io

### PR DESCRIPTION
**Pull request by stevepeak**
Uploading reports to [Codecov.io](https://codecov.io/)

Supported `json` format:
```json
{
  "uploader": "R",
  "files": [
    {
      "name": "string",
      "coverage": [null, 0, null, 1]
    }
  ]
}
```

> Branch coverage is supported via `"1/2"` for partial coverage or `"2/2"` for full branch coverage. This is optional if R can produce this data.

My first R lang experience. Help me fix any issues presented. Special thanks to @daroczig and @RomanTsegelskyi

### CODE REVIEW

File: R/codecov.R [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+    coverage_names,
+    sources,
+    coverages,
+    SIMPLIFY = FALSE,
+    USE.NAMES = FALSE)
+
+  jsonlite::toJSON(na = "null", list("files" = res, "uploader" = "R"))
+}
+
+per_line <- function(x) {
```
Comment (daroczig) [inline R/codecov.R]: You do not have to copy this function from the `coveralls.R` file, this function shares the same namespace.

Comment (stevepeak) [inline R/codecov.R]: Removed it. Thanks!

File: R/codecov.R [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+  # --------
+  # Codeship
+  # --------
+  } else if (Sys.getenv('CI') == "true" && Sys.getenv('CI_NAME') == 'codeship') {
+    # https://www.codeship.io/documentation/continuous-integration/set-environment-variables/
+    codecov_url <- paste(base_url, "/upload/v2?service=codeship",
+                         "&branch=", Sys.getenv('CI_BRANCH'),
+                         "&build=", Sys.getenv('CI_BUILD_NUMBER'),
+                         "&build_url=", Sys.getenv('CI_BUILD_URL'),
+                         "&commit=", Sys.getenv('CI_COMMIT_ID')), sep="")
```
Comment (daroczig) [inline R/codecov.R]: There's an extra closing parenthesis before the `sep` -- and the same issue also occurs at a few other lines below.

**Comment by daroczig** [comment]
I'd be very happy to see/help this working -- especially due to the number of supported CI backends, really impressive. Thank you @stevepeak!

File: R/codecov.R (hunk 2 above)
Comment (stevepeak) [inline R/codecov.R]: Fixed

**Comment by stevepeak** [comment]
Thank you! I'm really excited to see this through!

If you can send some tests though that would be great. A "fake" private repo token you can use for debugging is `473c8c5b-10ee-4d83-86c6-bfd72a185a27` you can read more here: https://codecov.io/api

I've already added support for this **specific** json layout.

**Comment by daroczig** [comment]
@stevepeak cool, I'll be able to run some tests tomorrow and will report back here

**Comment by jimhester** [comment]
you can ignore the absolute path lints by putting a `# nolint` comment after the lines

**Comment by jimhester** [comment]
First thank you for working on this, it is a major addition!

See my line notes for some small comments.

The last thing that would be nice is at least one test for each of the types similar to [test-coveralls.R](https://github.com/jimhester/covr/blob/12cefe4dc299bb643ec3bf602590e7aacb97cd6b/tests/testthat/test-coveralls.R).  However since you have limited experience with R I can tackle that unless you want to give it a go.

**Comment by stevepeak** [comment]
Your welcome! I look forward to having this accomplished. 

> Codecov requires a repo being backed by Github, Bitbucket or GitLab (coming soon). So running outside git is not supported.

I would ask for your help with writing tests. I appreciate you helping out! Let me know if there is any info you need.

File: R/codecov.R [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+#' @param path file path to the package
+#' @param Codecov url (change for Enterprise)
+#' @param ... additional arguments passed to \code{\link{package_coverage}}
+#' @export
+codecov <- function(path = ".", base_url = "https://codecov.io", ...) {
+  
+  # -------
+  # Jenkins
+  # -------
+  if (Sys.getenv("JENKINS_URL")) {
```
Comment (daroczig) [inline R/codecov.R]: `Sys.getenv("JENKINS_URL")` usually returns an empty string, which is not a valid logical expression to be passed to `if`. Maybe `Sys.getenv("JENKINS_URL") != ''`

File: R/codecov.R [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+    # path <- Sys.getenv("WORKSPACE")
+    codecov_url <- paste0(base_url, "/upload/v2?service=jenkins",
+                          "&branch=", Sys.getenv("GIT_BRANCH"),
+                          "&commit=", Sys.getenv("GIT_COMMIT"),
+                          "&build=", Sys.getenv("BUILD_NUMBER"),
+                          "&build_url=", Sys.getenv("BUILD_URL"))
+  # ---------
+  # Travis CI
+  # ---------
+  } else if (Sys.getenv("CI") == "true" && Sys.getenv("TRAVIS") == "true") {
```
Comment (daroczig) [inline R/codecov.R]: Not sure if it has to be in an `else` branch -- IMHO it should work fine (and indentation would be easier) with closing the previous `if` conditional and starting a new one

File: R/codecov.R [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+                          "&build=", Sys.getenv("WERCKER_MAIN_PIPELINE_STARTED"),
+                          "&owner=", Sys.getenv("WERCKER_GIT_OWNER"),
+                          "&repo=", Sys.getenv("WERCKER_GIT_REPOSITORY"),
+                          "&commit=", Sys.getenv("WERCKER_GIT_COMMIT"))
+  # ---------
+  # Local GIT
+  # ---------
+  } else {
+    branch <- local_branch()
+    codecov_url <- paste0(base_url, "/upload/v2",
```
Comment (daroczig) [inline R/codecov.R]: This returned a `Method Not Allowed` (405) error when I tried to upload (L119) the below report:

```r
> coverage
{"files":[{"name":"/home/daroczig/projects/r-test//R/r-test.R","coverage":[null,null,null,null,null,2,2,2,0,2]}],"uploader":["R"]} 
```

File: R/codecov.R (hunk 4 above)
Comment (jimhester) [inline R/codecov.R]: They do have to be `else if` if you want to check for a local git only if all the above conditions fail on [codecov.R#L101](https://github.com/stevepeak/covr/blob/master/R/codecov.R#L101)

**Comment by jimhester** [comment]
I should have mentioned this earlier, but the query parameters should not really be appended to the url like you are doing when using `httr`.

Rather you should construct a list of them and pass it to the `query` argument to `httr::POST`.

So instead of 

```r
    codecov_url <- paste0(base_url, "/upload/v2?service=travis-org",
                          "&branch=", Sys.getenv("TRAVIS_BRANCH"),
                          "&build=", Sys.getenv("TRAVIS_JOB_NUMBER"),
                          "&pull_request=", pr,
                          "&travis_job_id=", Sys.getenv("TRAVIS_JOB_ID"),
                          "&owner=", strsplit(Sys.getenv("TRAVIS_REPO_SLUG"), "/")[0],
                          "&repo=", strsplit(Sys.getenv("TRAVIS_REPO_SLUG"), "/")[1],
                          "&commit=", Sys.getenv("TRAVIS_COMMIT"))
```

You should do
```r
codecov_url <- paste0(base_url, "/upload/v2?service=travis-org")
codecov_query <- list(branch = Sys.getenv("TRAVIS_BRANCH"),
                      build = Sys.getenv("TRAVIS_JOB_NUMBER"),
                      pull_request = pr,
                      travis_job_id = Sys.getenv("TRAVIS_JOB_ID"),
                      owner = strsplit(Sys.getenv("TRAVIS_REPO_SLUG"), "/")[0],
                      repo = strsplit(Sys.getenv("TRAVIS_REPO_SLUG"), "/")[1],
                      commit = Sys.getenv("TRAVIS_COMMIT"))
```
then at [codecov.R#L119](https://github.com/stevepeak/covr/blob/ed1700ddcaaa97bbc7b86dc0675cbbc41bfb7871/R/codecov.R#L119) pass `codecov_query` as a parameter

```r
httr::content(httr::POST(url = codecov_url, query = codecov_query, body = list(json_file = httr::upload_file(name))))
```

File: R/codecov.R [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+      list("name" = jsonlite::unbox(name),
+        "coverage" = coverage)
+    },
+    coverage_names,
+    sources,
+    coverages,
+    SIMPLIFY = FALSE,
+    USE.NAMES = FALSE)
+
+  jsonlite::toJSON(na = "null", list("files" = res, "uploader" = "R"))
```
Comment (jimhester) [inline R/codecov.R]: The uploader should be unboxed

```r
jsonlite::toJSON(na = "null", list("files" = res, "uploader" = jsonlite::unbox("R")))
```

File: R/codecov.R (hunk 5 above)
Comment (stevepeak) [inline R/codecov.R]: This works below. Perhaps it was the '["R"]'?

```sh
curl -X POST https://codecov.io/upload/v2?token=473c8c5b-10ee-4d83-86c6-bfd72a185a27&branch=master&service=jenkins&commit=63f3e1944dbc020bae68f984d9fdd374806655a3 \
     -d '{"files":[{"name":"/home/daroczig/projects/r-test//R/r-test.R","coverage":[null,null,null,null,null,2,2,2,0,2]}],"uploader":"R"}'
```

File: R/codecov.R (hunk 6 above)
Comment (stevepeak) [inline R/codecov.R]: 6629384

File: R/codecov.R (hunk 4 above)
Comment (stevepeak) [inline R/codecov.R]: Is there a suggested change here? Thanks

File: R/codecov.R [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+  # Local GIT
+  # ---------
+  } else {
+    branch <- local_branch()
+    codecov_url <- paste0(base_url, "/upload/v2") # nolint
+    codecov_query <- list(branch = ifelse(branch == "HEAD", "master", branch),
+                          commit = trim(system("git rev-parse HEAD", intern = TRUE)))
+  }
+
+  if (Sys.getenv("CODECOV_TOKEN")) {
```
Comment (daroczig) [inline R/codecov.R]: This is `''` if not set, otherwise a string -- so you might compare it with an empty string here.

File: R/codecov.R [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+    codecov_query["token"] <- Sys.getenv("CODECOV_TOKEN")
+  }
+
+  coverage <- to_codecov(package_coverage(path, relative_path = TRUE, ...))
+
+  name <- tempfile()
```
Comment (daroczig) [inline R/codecov.R]: I think there is no need to write `coverage` to a file, but you should simply `POST` the JSON as is. In my local fork, removing L113-117 and changing L118 to the following worked fine:

```r
httr::content(httr::POST(url = codecov_url, query = codecov_query, body = coverage, encode = "json"))
```

File: R/codecov.R [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+  con <- file(name)
+  writeChar(con = con, coverage, eos = NULL)
+  close(con)
+  on.exit(unlink(name))
+  httr::content(httr::POST(url = codecov_url, query = codecov_query, body = list(json_file = httr::upload_file(name))))
+}
+
+to_codecov <- function(x) {
+  coverages <- per_line(x)
```
Comment (daroczig) [inline R/codecov.R]: It took a while to debug this, but it seems that the Codecov API expects an extra `NULL` or `NA` compared to Coverall, so the returned list by `per_line` should be tweaked with a leading `NA`. Quick and dirty workaround for this line:

```r
coverages <- lapply(covr:::per_line(package_coverage()), function(x) c(NA, x))
```

Now the two services are in sync:
* https://coveralls.io/builds/2105425/source?filename=.%2FR%2Fr-test.R
* https://codecov.io/github/daroczig/r-test/R/r-test.R?ref=master

**Comment by daroczig** [comment]
I've run a few tests on the current state of the pull request, and it seems to does its job. Looking forward to see it merged :)

**Comment by stevepeak** [comment]
Awesome! I would show the example on how to use this at [codecov/example-r](https://github.com/codecov/example-r). @daroczig can Travis test R yet? Could you help me prepare this repository?

**Comment by daroczig** [comment]
@stevepeak yeah, Travis can be used there -- if you add me as a collaborator, I can setup a minimal repo this weekend.

**Comment by stevepeak** [comment]
All set! Thank you very much :)

**Comment by daroczig** [comment]
Thanks, I've added a minimal pkg at https://github.com/codecov/example-r
And the first report should also soon appear at https://codecov.io/github/codecov/example-r

I will add a few other tests in separate commits to produce more reports and will also add a `README.md` in the next few days.

**Comment by daroczig** [comment]
I've tested this pull requested with Travis: https://codecov.io/github/codecov/example-r
Seems to be pretty stable now. @jimhester, what do you think?

**Comment by jimhester** [comment]
I think it is good! I added tests for the functions as well and fixed a few issues.

**Comment by stevepeak** [comment]
Nice job, thank you!

Could we add some copy on the readme too?

**Comment by daroczig** [comment]
Wow, thank you @jimhester!
@stevepeak did you already announce the new lang support?

**Comment by stevepeak** [comment]
Its on our Supported Languages https://codecov.io/#languages

I'll shoot out a tweet at 11am PST, I'll put the status id in this feed so you can interact. Thanks friends!

**Comment by daroczig** [comment]
@stevepeak the usage details are incorrect, please see https://github.com/codecov/example-r#-use-with-travis

**Comment by stevepeak** [comment]
Woops! Thanks, I'll adjust this promptly

**Pull request merged**